kernel/typing.ml

```diff
@@ -180,45 +230,75 @@ let rec pseudo_u sg (fail: int*term*term-> unit) (sigma:SS.t) : (int*term*term)
+          ( try sure_occur_check sg q (fun k -> k <= q) t; keepon ()
+            with VarSurelyOccurs -> warn () )
 
+        (* X = Y :  map either X to Y or Y to X *)
         | DB (l1,x1,n1), DB (l2,x2,n2) when n1>=q && n2>=q ->
            let (n,t) = if n1<n2
                        then (n1,mk_DB l2 x2 (n2-q))
                        else (n2,mk_DB l1 x1 (n1-q)) in
            pseudo_u sg fail (SS.add sigma (n-q) t) lst
+
+        (* X = t :
+           1) make sure that t is possibly closed
+           2) if by chance t already is closed, map X to t  *)
         | DB (_,_,n), t when n>=q ->
           begin
-            let n' = n-q in
-            match unshift_reduce sg q t with
-            | None -> warn ()
-            | Some ut ->
-               let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
-               if Subst.occurs n' t' then warn ()
-               else pseudo_u sg fail (SS.add sigma n' t') lst
+            try
+              sure_occur_check sg q (fun k -> k <= q || k = n) t;
+              match unshift_reduce sg q t with
+              | None -> keepon ()
+              | Some ut ->
+                let n' = n-q in
+                let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
+                if Subst.occurs n' t' then warn ()
+                else pseudo_u sg fail (SS.add sigma n' t') lst
+            with VarSurelyOccurs -> warn ()
           end
         | t, DB (_,_,n) when n>=q ->
           begin
-            let n' = n-q in
-            match unshift_reduce sg q t with
-            | None -> warn ()
-            | Some ut ->
-               let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
-               if Subst.occurs n' t' then warn ()
-               else pseudo_u sg fail (SS.add sigma n' t') lst
+            try
+              sure_occur_check sg q (fun k -> k <= q || k = n) t;
+              match unshift_reduce sg q t with
+              | None -> keepon ()
+              | Some ut ->
+                let n' = n-q in
+                let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
+                if Subst.occurs n' t' then warn ()
+                else pseudo_u sg fail (SS.add sigma n' t') lst
+            with VarSurelyOccurs -> warn ()
           end
 
-        | App (DB (_,_,n),_,_), _  when n >= q ->
-          if R.are_convertible sg t1' t2' then keepon () else warn ()
-        | _ , App (DB (_,_,n),_,_) when n >= q ->
-          if R.are_convertible sg t1' t2' then keepon () else warn ()
-
-        | App (Const (l,cst),_,_), _ when not (Signature.is_static sg l cst) -> keepon ()
-        | _, App (Const (l,cst),_,_) when not (Signature.is_static sg l cst) -> keepon ()
+        (* f t1 ... tn    /    X t1 ... tn  =  u
+           1) Gather all free variables in t1 ... tn
+           2) Make sure u only relies on these variables
+        *)
+        | App (DB (_,_,n),a,args), t when n >= q ->
+          let occs = Array.make q false in
```

Comment:
These two lines appears at least 4 times, can't you factorize them?

Reply:
There are also a block of 12 lines and another block of 2 lines duplicated above. That function duplicates lines because of symmetrical cases.

Reply:
But I have the impression from your code that the creation of the array could be done directly in the function gather_free_vars. Moreover, really seems copy/paste code which could be easily avoided. Maybe even the symmetrical cases could be handled in the same way.

What I am saying here is that maybe you could create a local function which handle these cases.